README.md
# gloTK

GloTK is the Genomes of Luminous Organisms Toolkit. It contains
scripts to facilitate genome assembly optomization and analysis using
common genome assemblers.

## Supported Assemblers

Currently there are only scripts for one assembler:

- Meraculous *2.2.2.4+*

## Dependencies

- The `mer_reporter.py` script currently requires the following to be installed:
  - Python 3.5
    - (If you haven't installed yet, I recommend the [Anaconda](https://www.continuum.io/downloads) distribution)
  - [py-gfm](https://py-gfm.readthedocs.io/en/latest/)
    - _Python Github-flavored Markdown_

## Install

You can install the `gloTK` package on your linux machine by typing
`pip install gloTK` in your terminal.

Major releases of `gloTK` are uploaded to the [Python Package Index](https://pypi.python.org/pypi/gloTK).

## Scripts

To use these programs, install `gloTK` as described above and type `glotk` into
your terminal. If you hit the `tab` key a few times your computer should try to
autocomplete and you will see a few options:

- **glotk-mer-reporter.py**
  - Generates HTML reports of all Meraculous assemblies found within
    the current working directory _(CWD)_. Collates all relevant files and images
    into a `meraculous_reports` directory in the _CWD_.
- **glotk-sweep**
  - This allows the user to input a single meraculous config file, and to define 
    parameters to sweep on and optomize. Currently this program supports sweeping 
    on `mer_size` and `bubble_depth_threshold`. Typing `glotk-sweep` into your
    terminal after installing will give you more details.

## News


### _*v0.1.15*_ - 20161116


* This version incorporates glotk-project. This script reads in a config file
  and generates a 'glotk_project' directory in which information about the
  libraries, assemblies, et cetera for this Meraculous genome project is stored. 
  Assemblies, reports, config files, et cetera  will take place in
  adjacent directories
* gloTK project files will now take on the following file structure:

```
project_dir/
|--glotk_info/
|  |  default_config.config
|  |  sample_metadata.config
|  |  read_metadata.config
|  |  assemblynumber_to_runname.txt
|  |
|  |--activity_log/
|  |     as000.log
|  |     as001.log
|  |     etcetera.log
|  |
|  |--read_configs/
|        reads0.yaml #this is the initial yaml file for imported reads
|        reads1.yaml #reads generated from Seqprep, trimmomatic, et cetera
|
|--glotk_assemblies
|  |--as000_<assembly_name>/
|  |  |--assembler_output/
|  |
|  |--as001_20161102_ME_pleu_bach_k63_d1/
|     |--meraculous_etcetera
|
|--glotk_configs
|     as000_<assembly_name>.config
|     as001_20161102_ME_pleu_bach_k63_d1.config
|
|--glotk_reads
|  |  reads.log
|  |
|  |--reads0/
|  |     <forward_reads_symlink>.fq.gz
|  |     <reverse_reads_symlink>.fq.gz
|  |
|  |--reads1/
|        <processed_forward>.fq.gz
|        <processed_reverse>.fq.gz
|
|--glotk_fastqc
|--glotk_kmers
|--glotk_reports
```


### _*v0.1.14*_ - 20161102


* Gave a --triplet flag to glotk-sweep to cause the assembler to run diploid modes
  0, 1, and 2 on the same kmer size.


### _*v0.1.13*_ - 20161102

* Made glotk-sweep function by inputting each value of k to assemble for, rather
  than using 'sweep start', 'sweep stop', and 'sweep interval'.
  Usage is: '--slist 23 27 57' to perform assemblies for those three values


NEWS
====================
 v0.1.15 - 20161116 
====================

* This version incorporates glotk-project. This script reads in a config file
  and generates a 'glotk_project' directory in which information about the
  libraries, assemblies, et cetera for this Meraculous genome project is stored. 
  Assemblies, reports, config files, et cetera  will take place in
  adjacent directories
* gloTK project files will now take on the following file structure:

```
project_dir/
|--glotk_info/
|  |  default_config.config
|  |  sample_metadata.config
|  |  read_metadata.config
|  |  assemblynumber_to_runname.txt
|  |
|  |--activity_log/
|  |     as000.log
|  |     as001.log
|  |     etcetera.log
|  |
|  |--read_configs/
|        reads0.yaml #this is the initial yaml file for imported reads
|        reads1.yaml #reads generated from Seqprep, trimmomatic, et cetera
|
|--glotk_assemblies
|  |--as000_<assembly_name>/
|  |  |--assembler_output/
|  |
|  |--as001_20161102_ME_pleu_bach_k63_d1/
|     |--meraculous_etcetera
|
|--glotk_configs
|     as000_<assembly_name>.config
|     as001_20161102_ME_pleu_bach_k63_d1.config
|
|--glotk_reads
|  |  reads.log
|  |
|  |--reads0/
|  |     <forward_reads_symlink>.fq.gz
|  |     <reverse_reads_symlink>.fq.gz
|  |
|  |--reads1/
|        <processed_forward>.fq.gz
|        <processed_reverse>.fq.gz
|
|--glotk_fastqc
|--glotk_kmers
|--glotk_reports
```

====================
 v0.1.14 - 20161102
====================

* Gave a --triplet flag to glotk-sweep to cause the assembler to run diploid modes
  0, 1, and 2 on the same kmer size.

====================
 v0.1.13 - 20161102
====================

* Made glotk-sweep function by inputting each value of k to assemble for, rather
  than using 'sweep start', 'sweep stop', and 'sweep interval'.
  Usage is: '--slist 23 27 57' to perform assemblies for those three values


=========
 v0.1.12
=========

* Fixed an error where not specifying the start index or assembly
  prefix caused the names to generate incorrectly or the program to
  crash.
* Added a test to verify that the assembler is working in sweep mode,
  including test data.


gloTK/scripts/glotk_sweep.py
#!/usr/bin/env python
# gloTK - Genomes of Luminous Organisms Toolkit
# Copyright (c) 2015-2016 Darrin Schultz. All rights reserved.
#
# This file is part of gloTK.
#
# GloTK is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# GloTK is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with GloTK.  If not, see <http://www.gnu.org/licenses/>.

"""

title: glotk-sweep
authr: darrin schultz

This program:
1. Reads in a meraculous config file and outputs all of the associated config
   files to $PWD/configs
2. The name of each run and the path to the directory is passed to a
   multiprocessing core that controls which assemblies are executed and when.
3. Each assembly is executed.

Usage:
"--slist 21 23 57 73" to perform assemblies for kmer sizes 21, 23, 57, 73, et cetera
"""

#import things for rest of program
import sys
import argparse
import os

#stuff for call shell commands
import subprocess

#multiprocessing stuff
from functools import partial
from multiprocessing import cpu_count
from multiprocessing import Pool
from multiprocessing.dummy import Pool as ThreadPool

#import gloTK stuff
from gloTK import MerParse
from gloTK import MerRunAnalyzer
import gloTK.utils

#This class is used in argparse to expand the ~. This avoids errors caused on
# some systems.
class FullPaths(argparse.Action):
    """Expand user- and relative-paths"""
    def __call__(self, parser, namespace, values, option_string=None):
        setattr(namespace, self.dest,
                os.path.abspath(os.path.expanduser(values)))

class CommandLine:
    """
    Handle the command line, usage and help requests.
    """
    def __init__(self) :
        """For stage 1, the necessary arguments for the MerParse class are:
          - inputFile
          - sweep
          - sList
          - asPrefix
          - asSI
          - genus
          - species
        """

        self.parser=argparse.ArgumentParser(description=__doc__,
                            formatter_class=argparse.RawDescriptionHelpFormatter)
        group = self.parser.add_mutually_exclusive_group(required=True)
        group.add_argument("-i", "--inputConfig",
                            type=str,
                            action=FullPaths,
                            help="""Uses a meraculous config file upon which to
                            base the following runs""")
        group.add_argument("-r", "--readsNum",
                            type = int,
                            help = """Bases the assembly on a yaml file for
                            reads that exist in this gloTK project directory.""")
        self.parser.add_argument("-s", "--sweep",
                            type=str,
                            choices=["mer_size", "bubble_depth_threshold"],
                            required=True,
                            help="""The parameter to sweep through in the
                            assembly.""")
        self.parser.add_argument("--slist",
                            type=str,
                            nargs='+',
                            help="""The values to sweep through""")
        self.parser.add_argument("-p", "--prefix",
                            type=str,
                            default='as',
                            help="""The assembly prefix used to name config files
                            and the output directories.""")
        self.parser.add_argument("-I", "--index",
                            type=str,
                            help="""The starting assembly index for naming the config
                            files and output directory.""")
        self.parser.add_argument("-G", "--genus",
                            type=str,
                            help="""The genus name for the sample that will be used
                            for naming config files and directory names.""")
        self.parser.add_argument("-S", "--species",
                            type=str,
                            help="""The species name for the sample that will be used
                            for naming config files and directory names.""")
        self.parser.add_argument("-n", "--simultaneous",
                            type=int,
                            default=1,
                            help="""The number of simultaneous assemblies to run.""")
        self.parser.add_argument("-M", "--maxProcs",
                            type=int,
                            default=cpu_count() - 2,
                            help="""The total number of processers used by all
                            of the assemblies combined.""")
        self.parser.add_argument("-c", "--censor",
                            type=str,
                            nargs='+',
                            help="""list of strings to censor for sending docs to
                            non-collaborators""")
        self.parser.add_argument("-q", "--quiet",
                            action='store_true')
        self.parser.add_argument("-t", "--triplet",
                            action='store_true',
                            help="""This performs diploid modes 0, 1, 2 for
                            every kmer size input""")
        self.parser.add_argument("-C", "--cleanup",
                            type=int,
                            default=1,
                            choices=[0,1,2],
                            help="""The cleanup level to pass along to the
                            run_meraculous.sh program.""")

    def parse(self):
        self.args = self.parser.parse_args()
        # Verify that we are in a gloTK directory
        setattr(self.args, "gloTKDir", os.getcwd())
        if not gloTK.utils.dir_is_glotk(self.args.gloTKDir):
            raise ValueError("""This is not a gloTK directory. Please initialize
            this directory with glotk-project.""")
        # 1. Reads in a meraculous config file and outputs all of the associated config
        #    files to $PWD/configs

        if self.args.inputConfig:
            setattr(self.args, "inputSource", self.args.inputConfig)
        elif self.args.readsNum != None:
            yamlPath = os.path.join(self.args.gloTKDir,
                                "gloTK_info/read_configs/reads{}.yaml".format(
                                    self.args.readsNum))
            setattr(self.args, "inputSource", yamlPath)
        print(self.args)
        if not os.path.exists(self.args.inputSource):
            raise IOError ("""You have selected an input source that does not
            exist: {}""".format(self.args.inputSource))

def mer_runner_dummy(instance):
    """This method is a helper method for class MerRunner. It allows
    multiprocessing module in python to map parallelism to a class method."""
    instance.meraculous_runner()

class MerRunner:
    """This class has one instance per Meraculous run and is accessed with the
    partial module"""
    def __init__(self, runName, configPath, cleanup, gloTKDir):
        """The cleanup parameter is what is passed to the run_meraculous script"""
        self.runName = runName
        self.configPath = configPath
        self.cleanup =  cleanup
        self.gloTKDir = gloTKDir
        self.allAssembliesDir = os.path.join(self.gloTKDir, "gloTK_assemblies")
        self.thisAssemblyDir = os.path.join(self.allAssembliesDir, self.runName)
        self.reportsDir = os.path.join(self.gloTKDir, "gloTK_reports")

        self.callString = "run_meraculous.sh -c {0} -dir {1} -cleanup_level {2}".format(
            self.configPath, self.runName, self.cleanup)

    def meraculous_runner(self):
        """
        Check to make sure that the allAssembliesDir has been created, if not,
        make it. This will only execute for the first time an assembly has been
        run in this directory.

        Run the directory from allAssembliesDir. The self.callString instance
        attribute tells Meraculous to name the assembly directory self.runName.

        After the run is complete, create the meraculous report, passing the
        directory containing the run (aka self.thisAssemblyDir).
        """
        #set the dir to temp assembly dir
        os.chdir(self.allAssembliesDir)

        print(self.callString)
        p = subprocess.run(self.callString, shell=True, stdout=subprocess.PIPE,
                           stderr=subprocess.PIPE,
                           universal_newlines=True)
        output = str(p.stdout)
        err = str(p.stderr)

        #generate the report for the run
        self._generate_report()

        #exit, returning the output and err
        return (output, err)

    def _generate_report(self):
        reporter = MerRunAnalyzer(self.thisAssemblyDir, self.gloTKDir, [])
        reporter.generate_report()

def main():
    """
    1. Reads in a meraculous config file and outputs all of the associated config
       files to $PWD/configs
    2. The name of each run and the path to the directory is passed to a
       multiprocessing core that controls which assemblies are executed and when.

    """
    parser = CommandLine()
    #this block from here: http://stackoverflow.com/a/4042861/5843327
    if len(sys.argv)==1:
        parser.parser.print_help()
        sys.exit(1)
    parser.parse()
    myArgs = parser.args
    print(myArgs)

    #Figure out how many processors to give to each assembly since we will be
    # running some things in parallel. The MerParse class will handle overriding
    # whatever is found in the config file in the read_config() method.
    procsPerAssembly = min(50, int(myArgs.maxProcs / myArgs.simultaneous))
    setattr(myArgs, "maxProcs", procsPerAssembly)


    merparser = MerParse(myArgs.inputSource,
                         gloTKDir = myArgs.gloTKDir,
                         sweep = myArgs.sweep,
                         sList = myArgs.slist,
                         lnProcs = myArgs.maxProcs,
                         asPrefix = myArgs.prefix,
                         asSI = myArgs.index,
                         genus = myArgs.genus,
                         species = myArgs.species,
                         triplet = myArgs.triplet)
    subParams = merparser.sweeper_output()
    configPaths = merparser.save_configs()
    print(configPaths)

    #make the assemblies dir ONCE to avoid a race condition for os.makedirs()
    allAssembliesDir = os.path.join(myArgs.gloTKDir, "gloTK_assemblies")
    if not os.path.exists(allAssembliesDir):
        os.makedirs(allAssembliesDir)

    #instantiate all of the classes that we will be using in parallel processing.
    # configPaths above returns a dict with the run name and abs path of config
    # as key:value pairs
    instances = []
    for runName in configPaths:
        configPath = configPaths.get(runName)
        #strip off the .config off the end of the runName, derived from configPath
        thisInstance = MerRunner(runName.strip(".config"), configPath,
                                 myArgs.cleanup, myArgs.gloTKDir)
        instances.append(thisInstance)

    if len(instances) == 0:
        print("There are no meraculous folders in this directory. Exiting")
    elif len(instances) > 0:
        print("Using {} processors.".format(procsPerAssembly * myArgs.simultaneous))
        # run the program for each instance
        # pool size is the number of simultaneous runs for the server
        pool = ThreadPool(myArgs.simultaneous)
        results = pool.map(mer_runner_dummy, instances)
        pool.close()
        pool.join()

if __name__ == "__main__":
    sys.exit(main())
